### Mục tiêu
Luyện tập xây dựng bảng Pivot table.

##### Bộ dữ liệu Online Retail chứa dữ liệu về giao dịch mua sắm online, các thông tin được thống kê bao gồm:

InvoiceNo: Số hóa đơn

StockCode: Mã kho

Description: Mô tả

Quantity: Số lượng

InvoiceDate: Ngày mua

UnitPrice: Giá

CustomerID: Mã khách hàng

Country: Quốc gia


### Yêu cầu
Đọc bộ dữ liệu, cho biết số dòng, số cột và kiểu dữ liệu của các thuộc tính.

In [1]:
# Nhập thư viện

import pandas as pd
import numpy as np

In [3]:
# Đọc dữ liệu

df = pd.read_csv('OnlineRetail.csv', encoding='unicode_escape')
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [4]:
# Xem số dòng và côt của dữ liệu

df.shape

(541909, 8)

In [5]:
# Xem kiểu dữ liệu của các cột (thuộc tính)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


#### Nhận xét:
 Description  540455 non-null  object 
 
 Quantity     541909 non-null  int64 

 --> cột 'Description' có giá trị khuyết thiếu

In [6]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

### Xây dựng bảng Pivot table, với mỗi Số hóa đơn tính trung bình cộng số lượng các mặt hàng theo từng Quốc gia. 


In [10]:
# bảng dữ liệu: df
# values: tính trung bình cộng số lượng các mặt hàng -> 'Quantity'
# index: mỗi Số hóa đơn... theo từng Quốc gia -> 'Country', 'InvoiceNo'
# dggfunc: mean

df1 = df.pivot_table(values='Quantity', index=['Country', 'InvoiceNo'], aggfunc='mean')
df1

Quantity
Country     InvoiceNo            
Australia   536389       7.642857
            537676      26.750000
            539419      14.600000
            540267     101.347826
            540280      70.000000
...                           ...
Unspecified 563947      18.000000
            564051       3.687500
            565303       3.909091
            576646       9.894737
            578539      13.323529

[25900 rows x 1 columns]

### Xây dựng bảng Pivot table, với mỗi Khách hàng cho biết số lượng mua hàng lớn nhất và nhỏ nhất theo Kho.


In [11]:
# với mỗi Khách hàng cho biết số lượng mua hàng lớn nhất và nhỏ nhất theo Kho.

#values: số lượng mua hàng -> 'Quantity'
# index: mỗi Khách hàng.. theo Kho -> CustomerID, StockCode
# aggfunc: min, max

df2 = df.pivot_table(values='Quantity', index=['CustomerID', 'StockCode'], aggfunc=['min','max'])
df2

min      max
                     Quantity Quantity
CustomerID StockCode                  
12346.0    23166       -74215    74215
12347.0    16008           24       24
           17021           36       36
           20665            6        6
           20719           10       10
...                       ...      ...
18287.0    84920            4        4
           85039A          48       48
           85039B          12       60
           85040A          12       36
           85173           48       48

[267615 rows x 2 columns]

### Xây dựng bảng Pivot table, với mỗi Mã kho tính tổng số lượng các mặt hàng và trung bình cộng giá.


In [12]:
# với mỗi Mã kho tính tổng số lượng các mặt hàng và trung bình cộng giá

# values:tổng số lượng các mặt hàng -> Quantity -> aggfunc= sum
# trung bình cộng giá -> UnitPrice -> addfunc: mean

#index: mỗi Mã kho -> StockCode

df3 = df.pivot_table(values=['Quantity', 'UnitPrice'], index='StockCode', aggfunc={'Quantity':'sum', 'UnitPrice':'mean'})
df3

Quantity  UnitPrice
StockCode                        
10002             1037   1.056849
10080              495   0.376667
10120              193   0.210000
10123C             -13   0.487500
10123G             -38   0.000000
...                ...        ...
gift_0001_20        20  15.038000
gift_0001_30        37  21.941250
gift_0001_40         3  33.566667
gift_0001_50         4  41.890000
m                    1   2.550000

[4070 rows x 2 columns]

### Xây dựng bảng Pivot table cho biết tổng số lượng hàng bán được của mỗi ngày.

In [13]:
# cho biết tổng số lượng hàng bán được của mỗi ngày

# values: Quantity	-> sum
# index: InvoiceDate

df4 = df.pivot_table(values='Quantity', index='InvoiceDate', aggfunc='sum')
df4

Quantity
InvoiceDate              
1/10/2011 10:04       -29
1/10/2011 10:07        -4
1/10/2011 10:08       -14
1/10/2011 10:32       260
1/10/2011 10:35       408
...                   ...
9/9/2011 8:48         104
9/9/2011 9:03         105
9/9/2011 9:13         372
9/9/2011 9:38         203
9/9/2011 9:52         782

[23260 rows x 1 columns]